File: osmToolsBasic/src/main/java/com/asamm/osmTools/cmdCommands/CmdContour.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package com.asamm.osmTools.cmdCommands;

import com.asamm.osmTools.Parameters;
import com.asamm.osmTools.mapConfig.ContourUnit;
import com.asamm.osmTools.mapConfig.ItemMap;

import java.io.File;
import java.io.IOException;

/**
 *
 * @author volda
 */
public class CmdContour extends Cmd{
    
    public CmdContour(ItemMap map) {
        super(map, ExternalApp.NO_EXTERNAL_APP);

        // check location of polygon
        if (!new File(map.getPathPolygon()).exists()){
            throw new IllegalArgumentException("Polygon file: " + map.getPathPolygon() + " does not exist");
        }

        // finally, set parameters
        addCommand(Parameters.phyghtDir);
    }
    
    public void createCmd() throws IOException {
        // prepare directory
        prepareDirectory(getMap().getPathContour());

        // add required commands
        addCommand("--polygon=" + getMap().getPathPolygon());
        if (getMap().getContourUnit() == ContourUnit.FEET){
            addCommand("--feet");
        }
        addCommand("--step=" + getMap().getContourStep());
        addCommand("--no-zero-contour");
        addCommand("--output-prefix=" + getMap().getPathContour());
        addCommand("--source=" + getMap().getContourSource()); // addCommand("--source=view3,view1");
        addCommand(getContourCatCmd());
        addCommand("--start-node-id=" + Parameters.contourNodeId);
        addCommand("--start-way-id=" + Parameters.contourWayId);
        addCommand("--write-timestamp");
        // add simplification of contour lines
        // for SRTM1 data, some value between 0.00003 and 0.00008 seems reasonable
        addCommand("--simplifyContoursEpsilon=0.00007");
        addCommand("--max-nodes-per-tile=0");
        addCommand("--hgtdir=" + Parameters.getHgtDir());

        int cores = Runtime.getRuntime().availableProcessors();
        addCommand("-j " + String.valueOf(cores)); //number of paralel jobs (POSIX only)
        if (Parameters.mapOutputFormat.endsWith("pbf")){
            //
            addCommand("--pbf");
        }
    }
    
    public void rename(){
        File ccDir = new File(getMap().getPathContour()).getParentFile();
        File files[];
        if (!ccDir.exists()){
            throw new IllegalArgumentException("Try to rename contour "+ getMap().getPathContour() +
                    ". Contour data dir does not exist");
        }
        files = ccDir.listFiles();
        
        for (int i = 0; i < files.length; i++){
            File file = files[i];
            if (file.isFile() && file.getName().startsWith(getMap().getName())){
                file.renameTo(new File(getMap().getPathContour()));
                break;
            }
        }
    }

    /**
     * Get command for major and minor category of contour lines.
     * Decide based on type of contour unit (meters or feet)
     */
    private String getContourCatCmd(){
        if (getMap().getContourUnit() == ContourUnit.FEET){
            return "--line-cat=400,200";
        }
        // for Meter and as fallback
        return "--line-cat=100,50";
    }
}

File: osmToolsBasic/src/main/java/com/asamm/osmTools/cmdCommands/CmdContour.kt
package com.asamm.osmTools.cmdCommands

import com.asamm.osmTools.config.AppConfig
import com.asamm.osmTools.mapConfig.ContourUnit
import com.asamm.osmTools.mapConfig.ItemMap
import com.asamm.osmTools.utils.Utils
import java.nio.file.Path
import kotlin.io.path.absolute

class CmdContour(val map: ItemMap) : Cmd(ExternalApp.PYHGTMAP) {

    private val TAG: String = CmdContour::class.java.simpleName

    private val tempMeter: Path = AppConfig.config.temporaryDir.resolve(AppConfig.config.contourConfig.tempMetersFile)
    private val tempFeet: Path = AppConfig.config.temporaryDir.resolve(AppConfig.config.contourConfig.tempFeetFile)

    init {
        require(map.isPlanet) {
            "Only planet map (id=planet) is supported for generation of contour lines"
        }
        require(AppConfig.config.contourConfig.polyCoverageMeter.toFile().exists()) {
            "Polygon file for meter contours: ${AppConfig.config.contourConfig.polyCoverageMeter.absolute()} does not exist"
        }
        require(AppConfig.config.contourConfig.polyCoverageFeet.toFile().exists()) {
            "Polygon file for feet contours: ${AppConfig.config.contourConfig.polyCoverageFeet.absolute()} does not exist"
        }
    }

    fun generate() {
        Utils.createParentDirs(tempMeter.toString())
        Utils.createParentDirs(tempFeet.toString())

        if (!tempFeet.toFile().exists()) {
            buildContoursCmd(ContourUnit.FEET, tempFeet).execute()
            rename(tempFeet)
        }

        if (!tempMeter.toFile().exists()) {
            buildContoursCmd(ContourUnit.METER, tempMeter).execute()
            rename(tempMeter)
        }

        Utils.createParentDirs(map.pathContour)
        CmdOsmium().merge(mutableListOf(tempMeter, tempFeet), map.pathContour)

        Utils.deleteFileQuietly(tempMeter)
        Utils.deleteFileQuietly(tempFeet)
    }

    private fun buildContoursCmd(unit: ContourUnit, outputPrefix: Path): ProcessCommand {
        val cores = Runtime.getRuntime().availableProcessors()
        return builder()
            .apply {
                when (unit) {
                    ContourUnit.METER -> {
                        add("--polygon=${AppConfig.config.contourConfig.polyCoverageMeter}")
                        add("--step=${AppConfig.config.contourConfig.stepMeter}")
                        add("--line-cat=${AppConfig.config.contourConfig.stepCategoryMeter}")
                        add("--start-node-id=${AppConfig.config.contourConfig.nodeIdMeter}")
                        add("--start-way-id=${AppConfig.config.contourConfig.wayIdMeter}")
                        add("--output-prefix=$outputPrefix")
                    }

                    ContourUnit.FEET -> {
                        add("--polygon=${AppConfig.config.contourConfig.polyCoverageFeet}")
                        add("--step=${AppConfig.config.contourConfig.stepFeet}")
                        add("--line-cat=${AppConfig.config.contourConfig.stepCategoryFeet}")
                        add("--start-node-id=${AppConfig.config.contourConfig.nodeIdFeet}")
                        add("--start-way-id=${AppConfig.config.contourConfig.wayIdFeet}")
                        add("--output-prefix=$outputPrefix")
                        add("--feet")
                    }

                    else -> throw IllegalArgumentException("Unsupported contour unit: $unit")
                }
            }
            .add("--no-zero-contour")
            .add("--source=${AppConfig.config.contourConfig.source}")
            .add("--write-timestamp")
            .add("--simplifyContoursEpsilon=0.00007")
            .add("--max-nodes-per-tile=0")
            .add("--hgtdir=${AppConfig.config.contourConfig.hgtDir}")
            .add("-j $cores")
            .add("--pbf")
            .build()
    }

    fun rename(tempPath: Path) {
        val ccDir = tempPath.parent.toFile()
        require(ccDir.exists()) {
            "Trying to rename contour $tempPath but parent directory does not exist"
        }
        val requiredFileName = tempPath.fileName.toString()
        val files = ccDir.listFiles() ?: return
        for (file in files) {
            if (file.isFile && file.name.startsWith(requiredFileName)) {
                file.renameTo(tempPath.toFile())
                break
            }
        }
    }
}
